### ISSUE TITLE
CR Snapshot Update Request for Web Neural Network API - webnn

### ISSUE THREAD

**Issue by dontcallmedom** [w3c/transitions#769]
# Document URLs
TR: https://www.w3.org/TR/webnn/
Current TR: https://www.w3.org/TR/2026/CRD-webnn-20260109/
Ed: https://webmachinelearning.github.io/webnn/

# Link to group's decision to request transition
https://www.w3.org/2026/01/15-webmachinelearning-minutes.html#f0f7

# Link to previous Candidate Recommendation transition or update request
https://github.com/w3c/transitions/issues/592

# Substantive changes
> The most notable additions include a third wave of operators for enhanced transformers support, the MLTensor API for buffer sharing, and a new abstract device selection mechanism. The API surface has been modernized and interoperability improvements have been made informed by wider implementation experience and developer feedback. This specification version strengthens security and privacy considerations including fingerprinting mitigations, and adds new accessibility considerations.

<details><summary>Detailed changes since <a href="https://www.w3.org/TR/2024/CR-webnn-20240411/">Candidate Recommendation Snapshot, 11 April 2024</a></summary>
New features ([class 4](https://www.w3.org/policies/process/#class-4))

* Expand operator set "wave 3" with new operators including dequantizeLinear, quantizeLinear, and attention operations ([#805](https://github.com/webmachinelearning/webnn/pull/805))
* Add MLTensor API, an interface for buffer sharing between WebNN and WebGPU, and reuse across multiple MLGraphs ([#787](https://github.com/webmachinelearning/webnn/pull/787))
* Add isNaN and isInfinite operators, new element-wise operators for checking NaN and infinite values ([#858](https://github.com/webmachinelearning/webnn/pull/858))
* Add roundEven operator, new rounding operator using banker’s rounding ([#859](https://github.com/webmachinelearning/webnn/pull/859))
* Add accelerator selection mechanism, a simple mechanism for selecting ML accelerators ([#895](https://github.com/webmachinelearning/webnn/pull/895))
* Expose WebNN API to shared workers and service workers to extend WebNN availability beyond window and dedicated worker ([#823](https://github.com/webmachinelearning/webnn/pull/823))
* Add optional operator labels for more diagnosable error messages ([#742](https://github.com/webmachinelearning/webnn/pull/742))
* Introduce MLNumber, a unified type for specifying numeric inputs of any type ([#647](https://github.com/webmachinelearning/webnn/pull/647))
* Add rankRange to opSupportLimits() to support specifying rank ranges in support limits ([#828](https://github.com/webmachinelearning/webnn/pull/828))
* Support rankRange for op output tensors to extend rankRange support to output tensors ([#857](https://github.com/webmachinelearning/webnn/pull/857))
* Add MLDeviceType npu, a Neural Processing Unit (NPU) device type ([#696](https://github.com/webmachinelearning/webnn/pull/696)) - this feature was added and removed since previous publication to simplify device selection, see [#809](https://github.com/webmachinelearning/webnn/pull/809)
* Add destroy() methods for MLContext and MLGraph, specify context loss behavior and error reporting ([#744](https://github.com/webmachinelearning/webnn/pull/744))
* Add an optional axis parameter to softmax operation ([#649](https://github.com/webmachinelearning/webnn/pull/649))
* Add outputDataType to argmin/argmax to support specifying output data type ([#730](https://github.com/webmachinelearning/webnn/pull/730))
* Allow arbitrary axes for resample2d to generalize resample2d to work with any axes ([#752](https://github.com/webmachinelearning/webnn/pull/752))
* Add Resample data type uint8/int8 to extend resample to support 8-bit integer types ([#891](https://github.com/webmachinelearning/webnn/pull/891))
* Simplify operand layout support of conv2d and pool2d operations, remove MLRoundingType from pool2d, simplify layout support ([#770](https://github.com/webmachinelearning/webnn/pull/770))
* Restrict padding options to better match backend limits ([#843](https://github.com/webmachinelearning/webnn/pull/843)) 

Other changes that do not add new features ([class 3](https://www.w3.org/policies/process/#class-3))

* Remove MLContext.compute() method in favor of MLTensor API ([#795](https://github.com/webmachinelearning/webnn/pull/795))
* Remove MLDeviceType to simplify device selection by removing device type enumeration ([#809](https://github.com/webmachinelearning/webnn/pull/809))
* Convert MLOperand methods to readonly attributes and change dataType() and shape() from methods to attributes ([#774](https://github.com/webmachinelearning/webnn/pull/774))
* Make MLOperandDescriptor.shape a required property ([#764](https://github.com/webmachinelearning/webnn/pull/764))
* Allow MLGraphBuilder.build() to be called only once to restrict build method to single invocation ([#717](https://github.com/webmachinelearning/webnn/pull/717))
* Remove fillSequence overload of constant() to remove sequence-based constant creation ([#656](https://github.com/webmachinelearning/webnn/pull/656))
* Swap parameters of scalar constant() operand method to reorder parameters for consistency ([#650](https://github.com/webmachinelearning/webnn/pull/650))
* Remove argmin/argmax selectLastIndex parameter to simplify argmin/argmax API ([#722](https://github.com/webmachinelearning/webnn/pull/722))
* Rename cast/constant parameter type to dataType for consistency across the API ([#888](https://github.com/webmachinelearning/webnn/pull/888))
* Replace MLActivation with MLRecurrentNetworkActivation for a more specific type for recurrent network activations ([#718](https://github.com/webmachinelearning/webnn/pull/718))
* Change DOMString to USVString for better Unicode support ([#715](https://github.com/webmachinelearning/webnn/pull/715))
* Rename parameter names of where to improve parameter naming for clarity ([#719](https://github.com/webmachinelearning/webnn/pull/719))
* Rename member output to outputs of MLLstmCellSupportLimits for consistency ([#757](https://github.com/webmachinelearning/webnn/pull/757))
* Reject promises for in-progress operations on destroyed MLTensor for proper promise rejection ([#799](https://github.com/webmachinelearning/webnn/pull/799))
* Specify operand data type constraints of operations to enable data type validation rules ([#646](https://github.com/webmachinelearning/webnn/pull/646))
* Add missing validation steps for several ops for validation improvements ([#820](https://github.com/webmachinelearning/webnn/pull/820))
* Add missing validation for pad(), slice(), and split() to enhance validation for array operations ([#690](https://github.com/webmachinelearning/webnn/pull/690))
* Simplify, correct, and add validation for GRU/LSTM for improved recurrent network validation ([#659](https://github.com/webmachinelearning/webnn/pull/659))
* Validate hidden size of GRU and LSTM operators ([#644](https://github.com/webmachinelearning/webnn/pull/644))
* Validate restriction of output padding in convTranspose2d for stricter validation for transposed convolution ([#631](https://github.com/webmachinelearning/webnn/pull/631))
* Enhance gather operation validation ([#642](https://github.com/webmachinelearning/webnn/pull/642))
* General validation improvements ([#643](https://github.com/webmachinelearning/webnn/pull/643))
* Improve resource validation ([#622](https://github.com/webmachinelearning/webnn/pull/622))
* Define error handling of MLNamedArrayBufferViews transfer algorithm to properly handle errors for buffer transfers ([#723](https://github.com/webmachinelearning/webnn/pull/723))
* Update dimension valid range to signed integer ([#738](https://github.com/webmachinelearning/webnn/pull/738))
* Introduce "valid dimension" concept to formalize dimension validity ([#641](https://github.com/webmachinelearning/webnn/pull/641))
* Ensure object creation specifies the realm for proper realm handling for object creation ([#810](https://github.com/webmachinelearning/webnn/pull/810))
* Clarify division operator rounding ([#909](https://github.com/webmachinelearning/webnn/pull/909))
* Correct pad scalar inconsistency ([#894](https://github.com/webmachinelearning/webnn/pull/894))
* Fix opSupportLimits error for split op ([#776](https://github.com/webmachinelearning/webnn/pull/776))
* Use EnforceRange for proper range enforcement in softmax() axis argument ([#746](https://github.com/webmachinelearning/webnn/pull/746))
* Add missing definitions of inputShape to conv2d algorithms ([#680](https://github.com/webmachinelearning/webnn/pull/680))
* Fix unidirectionally broadcast shapes steps ([#663](https://github.com/webmachinelearning/webnn/pull/663))
* Fix compute() promise rejection behavior if buffer transferring fails ([#639](https://github.com/webmachinelearning/webnn/pull/639))
* Update ArrayBufferView compatibility table with 64-bit integer types ([#698](https://github.com/webmachinelearning/webnn/pull/698)) 

Changes that do not functionally affect interpretation of the document ([class 2](https://www.w3.org/policies/process/#class-2))

* Make operand data type and rank validation table-driven for a systematic approach to validation ([#657](https://github.com/webmachinelearning/webnn/pull/657))
* Add non-normative table of operators by category ([#868](https://github.com/webmachinelearning/webnn/pull/868))
* Clarify cast() op behavior between different data types ([#726](https://github.com/webmachinelearning/webnn/pull/726))
* Clarify interpolation algorithms for resample2d ([#816](https://github.com/webmachinelearning/webnn/pull/816))
* Clarify reduction with empty axes and scalar input to clarify edge cases ([#741](https://github.com/webmachinelearning/webnn/pull/741))
* Add note about no-op graphs ([#665](https://github.com/webmachinelearning/webnn/pull/665))
* Add note about keepDimensions of reduction ops to clarify reduction operation behavior ([#648](https://github.com/webmachinelearning/webnn/pull/648))
* Reorganize emulation documentation ([#598](https://github.com/webmachinelearning/webnn/pull/598))
* Add decompositions for reduceLogSum, reduceLogSumExp, and reduceSumSquare ([#637](https://github.com/webmachinelearning/webnn/pull/637))
* Remove obsolete note about interop issues with clamp() minValue == maxValue ([#684](https://github.com/webmachinelearning/webnn/pull/684))
* Update specification boilerplate metadata information ([#769](https://github.com/webmachinelearning/webnn/pull/769)) 

Horizontal ([class 2](https://www.w3.org/policies/process/#class-2) and [class 3](https://www.w3.org/policies/process/#class-3))

* Add architectural resource contention considerations ([#765](https://github.com/webmachinelearning/webnn/pull/765))
* Add security considerations for Unicode ([#851](https://github.com/webmachinelearning/webnn/pull/851))
* Add security consideration for computation control-flow attacks ([#725](https://github.com/webmachinelearning/webnn/pull/725))
* Revise privacy considerations ([#890](https://github.com/webmachinelearning/webnn/pull/890))
* Add privacy considerations for opSupportLimits() fingerprinting ([#881](https://github.com/webmachinelearning/webnn/pull/881))
* Add accessibility considerations ([#869](https://github.com/webmachinelearning/webnn/pull/869))
* Add internationalization note regarding label usage ([#841](https://github.com/webmachinelearning/webnn/pull/841)) 

Editorial ([class 2](https://www.w3.org/policies/process/#class-2))

* Various editorial improvements ([#834](https://github.com/webmachinelearning/webnn/pull/834))
* Various style and wording tweaks ([#797](https://github.com/webmachinelearning/webnn/pull/797))
* Grammar and spelling corrections ([#782](https://github.com/webmachinelearning/webnn/pull/782))
* Simplify specification steps with helpers algorithms ([#737](https://github.com/webmachinelearning/webnn/pull/737))
* Improve type references ([#735](https://github.com/webmachinelearning/webnn/pull/735))
* Reference WebIDL transferable definition ([#732](https://github.com/webmachinelearning/webnn/pull/732))
* Avoid "sequence" in prose for improved terminology usage ([#729](https://github.com/webmachinelearning/webnn/pull/729))
* Improve logic for validating algorithm steps for enhanced linting and validation ([#727](https://github.com/webmachinelearning/webnn/pull/727))
* Link method argument definitions for Improved cross-references ([#721](https://github.com/webmachinelearning/webnn/pull/721))
* Remove unnecessary subsections for improved organization ([#711](https://github.com/webmachinelearning/webnn/pull/711))
* Link to "is empty" not "empty" when adjective intended ([#708](https://github.com/webmachinelearning/webnn/pull/708))
* Add utilities to ease authoring and review of this specification ([#702](https://github.com/webmachinelearning/webnn/pull/702))
* Fix "transferred" cross-reference ([#679](https://github.com/webmachinelearning/webnn/pull/679))
* Make "generically emulated" text a macro to improve authoring experience ([#638](https://github.com/webmachinelearning/webnn/pull/638))
* Fix emulation error of LSTM by 'backward' and 'both' directions ([#802](https://github.com/webmachinelearning/webnn/pull/802))
* Fix emulation error of GRU by 'backward' and 'both' directions ([#803](https://github.com/webmachinelearning/webnn/pull/803))
* Correct typos/JS errors in decompositions ([#699](https://github.com/webmachinelearning/webnn/pull/699))
</details>

# Any changes in requirements?
N/A

# Wide Review of substantive changes
Documented in https://github.com/webmachinelearning/webnn/issues/239#issuecomment-2740740891

# Issues status
https://github.com/webmachinelearning/webnn/issues

# Formal Objections
None

# Any changes in implementation information?
The WG is aware of interest for an origin trial in Chromium browsers, and has received a positive standards position from Mozilla. Apple is active in the group.

# Deadline for further comments
2 months after publication

# Any changes in patent disclosures?
None

NOTE: For more information, see [Organize a Technical Report Transition](https://www.w3.org/Guide/transitions?profile=CR&cr=snapshot).

cc @anssiko

**Comment by plehegar** [comment]
Transition approved.

And well done for your work around [wide review](https://github.com/webmachinelearning/webnn/issues/239). Much appreciated.

**Comment by xueyuanjia** [comment]
draft transition received.
